### ISSUE TITLE
CSS corner-shape

### ISSUE THREAD

**Issue by jsnkuhn** [mozilla/standards-positions#823]
## Request for Mozilla Position on an Emerging Web Specification

* Specification Title: CSS Backgrounds and Borders Module Level 4: Corner Shaping: the [corner-shape]
* Specification or proposal URL (if available): https://drafts.csswg.org/css-backgrounds-4/#corner-shaping 
* Explainer URL (if available): https://lea.verou.me/2013/03/preview-border-corner-shape-before-implementations/
* Caniuse.com URL (optional): 
* Bugzilla URL (optional): https://bugzilla.mozilla.org/show_bug.cgi?id=1726232
* Mozillians who can provide input (optional):

**Comment by zcorpan** [comment]
cc @dholbert

**Comment by dholbert** [comment]
I still need to review this, but in the meantime: the spec link has changed -- [it moved](https://github.com/w3c/csswg-drafts/pull/9048) here:
https://drafts.csswg.org/css-borders-4/#corner-shaping

**Comment by dholbert** [comment]
I think we should take a "positive" position here. The specced feature is fairly limited right now (just corner-shape:angle essentially), but the use-cases for that (cut corners, arrows) seem valid and are poorly-addressed by current web features. The ergonomics, implementability, and extensibility of this approach seem good as well.

**Comment by emilio** [comment]
@dholbert fwiw, I'm not so sure about it. `corner-shape: angle` messing with `border-radius` feels wrong somehow (even if it's the simplest way to specify it).

It feels like something like a way to make a clip-path also change the border shape might be a more generic way of solving these use cases? I'm pretty sure I've made the same comment in this repo somewhere to that effect but I forget where (is this issue a dup?)

**Comment by dholbert** [comment]
> @dholbert fwiw, I'm not so sure about it. `corner-shape: angle` messing with `border-radius` feels wrong somehow (even if it's the simplest way to specify it).

Yeah, the "radius" terminology is a bit odd for non-circular corners.  Essentially, I think the idea with this `corner-shape` generalization is just that `border-radius` tells you how far along each side to extend the corner decoration (whatever it is) before you reach a plain-vanilla border.

The naming could perhaps be clearer if we had a time machine and could redesign `border-radius` before it shipped. :)

> It feels like something like a way to make a clip-path also change the border shape might be a more generic way of solving these use cases? I'm pretty sure I've made the same comment in this repo somewhere to that effect but I forget where (is this issue a dup?)

That's definitely more generic, conceptually, but it might be clunkier for authors to use.
- they may not want to define a whole clip-path
- they may not want the content to be clipped (if they just want an arrow-shaped button, maybe with some sort of tooltip or decorator that extends outside of it that they don't want to clip)
- if they're applying this special shape only some of the time, and they have to do it using a clip-path-defined border, I worry that the clip-path-defined-border might make subtly different pixel-snapping decisions for flat border sides (either due to author bugs or browser bugs).  E.g. if you have a rectangular button and you want to add a hover style that cuts off the corners, and you have to define a clip path to specify the cut-off-corner button shape, then maybe that clip path ends up drawing the top/right/bottom/left edges at a slightly different position.

Maybe these are non-issues with what you've actually got in mind, though. I'm curious to see more details about how this might work. But I am drawn to the simplicity of just (ab)using border-radius to say how far the corner-decoration extends inwards on each side (which it already means, but you can only use it for circular/elliptical corner effects).

**Comment by o-t-w** [comment]
`clip-path` hides `box-shadow`. Developers will want to have elements with both fancy corners and shadows.

**Comment by SebastianZ** [comment]
> It feels like something like a way to make a clip-path also change the border shape might be a more generic way of solving these use cases? I'm pretty sure I've made the same comment in this repo somewhere to that effect but I forget where (is this issue a dup?)

> `clip-path` hides `box-shadow`. Developers will want to have elements with both fancy corners and shadows.

For what it's worth, that's discussed in https://github.com/w3c/csswg-drafts/issues/5881 and https://github.com/w3c/csswg-drafts/issues/6997. Point there is to allow changing an element's shape in a way that borders and box shadows are affected as well.

I agree that allowing to change the element's shape via `clip-path` or something similar is the more powerful and flexible solution to this. Though it is also the more complex, at least from an author's perspective, but I assume also from an implementer's perspective. And many use cases will already be covered by the simpler solution of allowing different corner shapes.

And I also agree that the name `border-radius` feels a bit weird for that, though it can easily be explained how it interacts with different shapes. (Disregarding that `border-*` was always a misnomer for that property.)
(Though I'd be happy to introduce an alias for `border-radius` if that makes things feel more correct and easier to grasp. 😎)

Sebastian

**Comment by yisibl** [comment]
Chrome and Safari have already started to implement corner-shape, hopefully Firefox will join the discussion and implement it soon.

https://chromium-review.googlesource.com/c/chromium/src/+/6286466
https://github.com/WebKit/WebKit/pull/41078

**Comment by benface** [comment]
> Though I'd be happy to introduce an alias for border-radius if that makes things feel more correct and easier to grasp.

I think that would be ideal. Something like `corner-size` maybe?

**Comment by jsnkuhn** [comment]
Another issue with the reuse of `border-radius` for defining corner sizes is that it forces a rounded corner fallback for browsers that don't support "corner-shape". This is probably fine if you are using `corner-shape:  round` or `corner-shape: squircle` but it's not a good fallback for `angle/bevel`, `notch`, `scoop` , `square/straight` or any combination of these. Granted this can be worked around with `@supports` but it seems to be creating extra work for devs.

**Comment by noamr** [comment]
The `corner-shape` spec has changed significantly since originally starting this thread.
`corner-shape` is now a value that can go between fully concave (notch) and fully convex (square) and animate between these values using a `superellipse()` value. See https://drafts.csswg.org/css-borders-4/#corner-shaping. Would love to hear Mozilla's view on this!

**Comment by mr-cheffy** [comment]
Thought its worth mentioning that it has already been implemented in [Chrome 139](https://groups.google.com/a/chromium.org/g/blink-dev/c/OoX4xjqSPt4/m/m-DV3dV0BAAJ) for a while and its usage keeps increasing amongst more and more sites.

**Comment by Wolfsong** [comment]
That would be great to see this sooner than later! We have product, that actively moving to the new tailwind based cyberpunk ui, and cut corners via corner-shape - is crucial for transitioning and composability with other utility classes, while firefox being widely used among target audience.

**Comment by mr-cheffy** [comment]
I started developing it for firefox, `layout.css.corner-shape.enabled` should appear on nightly soon!

**Comment by yisibl** [comment]
> I started developing it for firefox, `layout.css.corner-shape.enabled` should appear on nightly soon!

I'm really looking forward to it.

**Comment by xblurx** [comment]
> I started developing it for firefox, `layout.css.corner-shape.enabled` should appear on nightly soon!

let's go!

**Comment by AzizStark** [comment]
> I started developing it for firefox, `layout.css.corner-shape.enabled` should appear on nightly soon!

can't wait for this!

**Comment by SebastianZ** [comment]
Already works, only borders are currently still missing. Thank you for working on this, @mr-cheffy!

Sebastian

**Comment by SebastianZ** [comment]
Since the implementation is already pretty complete now, I think it's fair to mark this as "position: positive".

Sebastian
